modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/TableRowConverterImpl.java

```diff
@@ -56,15 +74,41 @@ public class TableRowConverterImpl implements TableRowConverter {
 
         int size = requiredColumns == null ? descriptor.columnsCount() : requiredColumns.cardinality();
 
-        mapping = new int[size];
+        requiredColumnsMapping = new int[size];
+        fullMapping = new int[descriptor.columnsCount()];
+        keyMapping = new int[schemaDescriptor.keyColumns().length()];
+
+        for (int i = 0, j = 0, k = 0; i < descriptor.columnsCount(); i++) {
+            ColumnDescriptor column = descriptor.columnDescriptor(i);
+            int schemaIndex = schemaDescriptor.column(column.name()).schemaIndex();
+
+            fullMapping[schemaIndex] = i;
 
-        int currentIdx = 0;
-        for (ColumnDescriptor column : descriptor) {
-            if (requiredColumns != null && !requiredColumns.get(column.logicalIndex())) {
-                continue;
+            if (column.key()) {
+                keyMapping[j++] = schemaIndex;
             }
 
-            mapping[currentIdx++] = schemaDescriptor.column(column.name()).schemaIndex();
+            if (requiredColumns == null || requiredColumns.get(column.logicalIndex())) {
+                requiredColumnsMapping[k++] = schemaIndex;
+            }
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public <RowT> BinaryRowEx toBinaryRow(ExecutionContext<RowT> ectx, RowT row, boolean key) {
+        BinaryTuple binaryTuple = ectx.rowHandler().toBinaryTuple(row);
+
+        if (!key) {
+            FormatAwareProjectedTuple tuple = new FormatAwareProjectedTuple(binaryTuple, fullMapping);
+            BinaryRowImpl binaryRow = new BinaryRowImpl(schemaDescriptor.version(), tuple.byteBuffer());
+
+            return Row.wrapBinaryRow(schemaDescriptor, binaryRow);
```

Comment:
it's better to avoid using `Row` in this case. Both `wrapBinaryRow` and `wrapKeyOnlyBinaryRow` converts given schema descriptor to BinaryTupleSchema. For _every_ row. But, in fact, we don't need this tuple schema, because BinaryRow assumes the schema is known in advance. I think, it's better to introduce our own implementation of BinaryRowEx interface

Reply:
I agree. Fixed. 
But results are underwhelming and there is no noticeable improvement. 

>MAIN
>
>Benchmark                                                   (clusterSize)  (fsync)  (partitionCount)  Mode  Cnt       Score       Error   Units
>InsertBenchmark.sqlInsert                                               1    false                 1  avgt   60       0.253 ±     0.005   ms/op
>InsertBenchmark.sqlInsert:·gc.alloc.rate                                1    false                 1  avgt   60     324.030 ±    33.318  MB/sec
>InsertBenchmark.sqlInsert:·gc.alloc.rate.norm                           1    false                 1  avgt   60  107490.553 ± 10795.221    B/op
>InsertBenchmark.sqlInsert:·gc.churn.G1_Eden_Space                       1    false                 1  avgt   60     336.300 ±    26.883  MB/sec
>InsertBenchmark.sqlInsert:·gc.churn.G1_Eden_Space.norm                  1    false                 1  avgt   60  112959.135 ±  7889.125    B/op
>InsertBenchmark.sqlInsert:·gc.churn.G1_Old_Gen                          1    false                 1  avgt   60       0.001 ±     0.001  MB/sec
>InsertBenchmark.sqlInsert:·gc.churn.G1_Old_Gen.norm                     1    false                 1  avgt   60       0.179 ±     0.452    B/op
>InsertBenchmark.sqlInsert:·gc.churn.G1_Survivor_Space                   1    false                 1  avgt   60       0.306 ±     0.219  MB/sec
>InsertBenchmark.sqlInsert:·gc.churn.G1_Survivor_Space.norm              1    false                 1  avgt   60     102.044 ±    73.259    B/op
>InsertBenchmark.sqlInsert:·gc.count                                     1    false                 1  avgt   60     130.000              counts
>InsertBenchmark.sqlInsert:·gc.time                                      1    false                 1  avgt   60    1040.000                  ms
>
>PATCH
>
>Benchmark                                                   (clusterSize)  (fsync)  (partitionCount)  Mode  Cnt       Score       Error   Units
>InsertBenchmark.sqlInsert                                               1    false                 1  avgt   60       0.242 ±     0.009   ms/op
>InsertBenchmark.sqlInsert:·gc.alloc.rate                                1    false                 1  avgt   60     322.055 ±    34.048  MB/sec
>InsertBenchmark.sqlInsert:·gc.alloc.rate.norm                           1    false                 1  avgt   60  102107.568 ± 10207.842    B/op
>InsertBenchmark.sqlInsert:·gc.churn.G1_Eden_Space                       1    false                 1  avgt   60     339.585 ±    27.276  MB/sec
>InsertBenchmark.sqlInsert:·gc.churn.G1_Eden_Space.norm                  1    false                 1  avgt   60  109706.270 ±  8676.535    B/op
>InsertBenchmark.sqlInsert:·gc.churn.G1_Old_Gen                          1    false                 1  avgt   60       0.022 ±     0.043  MB/sec
>InsertBenchmark.sqlInsert:·gc.churn.G1_Old_Gen.norm                     1    false                 1  avgt   60       7.143 ±    13.789    B/op
>InsertBenchmark.sqlInsert:·gc.churn.G1_Survivor_Space                   1    false                 1  avgt   60       0.470 ±     0.299  MB/sec
>InsertBenchmark.sqlInsert:·gc.churn.G1_Survivor_Space.norm              1    false                 1  avgt   60     163.871 ±   113.886    B/op
>InsertBenchmark.sqlInsert:·gc.count                                     1    false                 1  avgt   60     136.000              counts
>InsertBenchmark.sqlInsert:·gc.time                                      1    false                 1  avgt   60    1108.000                  ms